### ISSUE TITLE
S3M: It's possible to tell GUS and SB apart!

### ISSUE THREAD

**Issue by sagamusix** [libxmp/libxmp#357]
Today I made an amazing discovery while looking at some S3Ms in a hex editor: It is in fact possible to tell whether an S3M file was last saved with the GUS driver loaded in ST3, or with the SB driver: In the sample headers, there are some reserved / internal values, in particular the `Int:Gp` field that contains the sample address in GUS memory.

- If the file was last saved with a SB driver loaded, this value is 0 for empty sample slots and 1 otherwise.
- If the file was last saved with the GUS driver loaded, this value is 0 for empty sample slots, and for other sample slots it's some value greater than or equal to 1 (the lowest address).
- Early versions of ScreamTracker 3.00 didn't have GUS support, so in S3M files saved with those versions all of those `Int:Gp` values are 0, which is okay because that would indicate a SoundBlaster driver anyway (e.g. party.s3m by Skaven - which was used in The Party '92 Invitation which indeed only supports SB output as well). However there are also *some* ScreamTracker 3.00 modules with the correct GUS pointer values in them, so this feature was added before the ScreamTracker 3.01 release.

Luckily, re-saving the same file in different ST3 sessions with GUS or SB loaded updates to values properly, so if you first save a file with the GUS driver loaded you get the GUS pointers in the file, and if you later resave it with the SB driver loaded all those pointer values become 0 and 1 again. So you get a pretty reliable fingerprint of whether the module is inteded to use GUS quirks or SB quirks.

As I recall, xmp already supports some GUS-specific quirks for S3M. By applying this fingerprint detection, those quirks could be disabled for S3Ms that were detected to be saved with the SB drivers loaded.

**Comment by sagamusix** [comment]
Note that the `Int:Gp` value being 0 for all samples despite sample data being present can also be used for fingerprinting other software disguising as ST3.20 (but not for ST3.00 due to early versions lacking GUS support, as explained above). For example there is some PSM to S3M conversion tool that otherwise disguises perfectly as ST3, but can still be identified due to `Int:Gp` being 0 for all samples.

**Comment by AliceLR** [comment]
Slightly relatedly, I updated my tools to dump S3Ms, and I noticed classic MPT modules seem to save with `Cwt/v`=0x1320 and `Int:Gp`=0 for all instruments—is this reliable fingerprinting? (libxmp currently doesn't try to detect classic MPT, but it seems like it should all things considered.)

**Comment by AliceLR** [comment]
Some fairly pointless data since I can aggregate it now (ModLand 2014 torrent):

```
Scream Tracker 3:	8277
Impulse Tracker:	1366
Modplug:		578
IMAGO Orpheus:		20
OpenMPT:		7
0x0208:			6
Schism Tracker:		2

ST3, Int:GP is SB:	4633
ST3, Int:GP is GUS:	3601
ST3, Int:GP is ???:	43
AdLib (inst only):	3
AdLib (inst and chn):	2
```

The 43 ST3-identified modules that failed to fingerprint as either using your criteria have all 0s in `Int:Gp` and all report `0x1300` or `0x1301`.

**Comment by sagamusix** [comment]
Those two points are not enough to identify MPT (that would also catch some other clones and conversion tools). You can look at OpenMPT's or Schism's code for more reliable MPT fingerprinting.

**Comment by sagamusix** [comment]
Also note that it's not possible to tell MPT Classic and OpenMPT 1.17 apart. OpenMPT only starts using its own tracker ID from 1.18 onwards.

**Comment by sagamusix** [comment]
Can you maybe point out some of those files which have all-zero  `Int:Gp` with tracker version `0x1301`? That combination doesn't seem to make any sense to me, and I don't seem to have any such files locally (apart from those that have no sample data at all, e.g. OPL-only S3Ms). For version `0x1300` the meaning is well-defined: It's an early ST3.00 version without GUS support (so implicitly SoundBlaster).

**Comment by AliceLR** [comment]
Sure. There aren't very many, so here's all of them:

[Inspekdah Deck/(c)94 the expressway.s3m](http://modland.com/pub/modules/Screamtracker%203/Inspekdah%20Deck/(c)94%20the%20expressway.s3m)
[Most (6) of the S3Ms on ModLand by Maelcum.](http://modland.com/pub/modules/Screamtracker%203/Maelcum/) (The exception is "Out Of Face".)
[All (4) of the S3Ms on ModLand by The Hacker.](http://modland.com/pub/modules/Screamtracker%203/The%20Hacker/)

I updated the detection to 1) detect `0x1300` with all 0 `Int:Gp` as SoundBlaster and 2) fix a bug that caused some others to false positive :(. These 11 modules are the only remaining "ST3" S3Ms in the modules I'm testing on with no identified sound card.

**Comment by sagamusix** [comment]
Hmm, in underground remix.s3m, there's a note saying "...even though I don't have an SB Pro", which probably means that it was written with the regular (not pro) SoundBlaster. I'm not aware of any other early trackers disguising as ST3.01. Creating a file in ST3.01 with no driver (e.g. by providing a bogus base address) also doesn't result in this situation (the file is detected as SB).

**Comment by sagamusix** [comment]
Oh, and those `0x0208` (I assume that's the raw cwtv value?) would also be interesting to look into. Kinda looks like an Impulse Tracker version with the tracker ID nibble missing...

**Comment by AliceLR** [comment]
> Oh, and those `0x0208` (I assume that's the raw cwtv value?) would also be interesting to look into. Kinda looks like an Impulse Tracker version with the tracker ID nibble missing...

Yes, that's just the raw cwtv value. [All of them are modules by Yerzmyey.](http://modland.com/pub/modules/Screamtracker%203/Yerzmyey/)

**Comment by sagamusix** [comment]
For posterity, the files in https://github.com/libxmp/libxmp/issues/357#issuecomment-933013473 were made with Zab's [To-S3M](https://demozoo.org/productions/329258/) converter, originally they were MTM files.

Other tools disguising as ST3.01 I have found are UNMO3 and deMODifier, a converter for SoundSmith files. Both are fairly easy ot fingerprint (can be found in OpenMPT source).

There's one more S3M file identifying as ST3.01 on ModLand that I can't fingerprint: [zeromus.s3m](https://files.exotica.org.uk/modland/?file=pub/modules/Screamtracker%203/-%20unknown/zeromus.s3m) - probably some sort of automatic SNES conversion.

**Comment by AliceLR** [comment]
Moving to 4.6.4 for the remaining triage at minimum (mainly need to figure out whatever GUS features libxmp actually does support for S3M).
